sql/expression/function/json_contains.go

```diff
@@ -41,18 +44,131 @@ import (
 // Otherwise, the candidate value is not contained in the target document.
 //
 // https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-contains
+// TODO: Add multi index optimization -> https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-multi-valued
 type JSONContains struct {
-	sql.Expression
+	JSONTarget    sql.Expression
+	JSONCandidate sql.Expression
+	Path          sql.Expression
 }
 
-var _ sql.FunctionExpression = JSONContains{}
+var _ sql.FunctionExpression = (*JSONContains)(nil)
 
 // NewJSONContains creates a new JSONContains function.
 func NewJSONContains(args ...sql.Expression) (sql.Expression, error) {
-	return nil, ErrUnsupportedJSONFunction.New(JSONContains{}.FunctionName())
+	if len(args) < 2 || len(args) > 3 {
+		return nil, sql.ErrInvalidArgumentNumber.New("JSON_CONTAINS", "2 or 3", len(args))
+	}
+
+	if len(args) == 2 {
+		return &JSONContains{args[0], args[1], nil}, nil
+	}
+
+	return &JSONContains{args[0], args[1], args[2]}, nil
 }
 
 // FunctionName implements sql.FunctionExpression
 func (j JSONContains) FunctionName() string {
 	return "json_contains"
 }
+
+func (j JSONContains) Resolved() bool {
```

Comment:
All of these need to be pointer receivers. If you're checking that you're implementing an interface via a pointer (`var _ sql.FunctionExpression = (*JSONContains)(nil)` here) and returning a pointer from `New`, then all of your functions should be pointers as well.

Reply:
Totally my bad. Will be more cautious next time